File: 2025/python/junior/junior5.py
r = int(input())
c = int(input())
m = int(input())

row = [m]
tile = (r - 1) * c % m + 1
for _ in range(c):
    row.append(tile)
    tile = tile % m + 1
row.append(m)

for i in range(r - 2, -1, -1):
    next_row = [m * (r - i)]
    for j in range(c):
        mm = min(row[j], row[j + 1], row[j + 2])
        next_row.append(mm + (i * c + j) % m + 1)
    next_row.append(m * (r - i))
    row = next_row

print(min(row))

File: 2025/cpp/junior/junior5.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  int r, c, m;
  cin >> r >> c >> m;

  vector<int> row{};
  int tile = (r - 1) * c % m + 1;
  for (int i = 0; i < c; i++) {
    row.push_back(tile);
    tile = tile % m + 1;
  }

  for (int i = r - 2; i >= 0; i--) {
    vector<int> nextRow{};
    for (int j = 0; j < c; j++) {
      int smallest = row[j];
      if (j > 0) {
        smallest = min(smallest, row[j - 1]);
      }
      if (j < c - 1) {
        smallest = min(smallest, row[j + 1]);
      }
      nextRow.push_back(smallest + (i * c + j) % m + 1);
    }
    row = nextRow;
  }

  cout << *min_element(row.begin(), row.end());
}
